src/Talis/Persona/Client/Tokens.php

```diff
@@ -56,51 +56,81 @@ public function validateToken($params = [])
-                $decoded = (array)JWT::decode($token, $cert, ['RS256']);
-            } else {
-                throw new \InvalidArgumentException('cannot parse public key');
-            }
-        } catch (\DomainException $exception) {
-            $this->getLogger()->error('Invalid signature', [$exception]);
-            return ValidationResults::InvalidSignature;
-        } catch (\InvalidArgumentException $exception) {
-            $this->getLogger()->error('Invalid public key', [$exception]);
-            return ValidationResults::InvalidPublicKey;
-        } catch (\UnexpectedValueException $exception) {
-            // Expired, before valid, invalid json, etc
-            $this->getLogger()->debug('Invalid token', [$exception]);
-            return ValidationResults::InvalidToken;
+            $decodedToken = $this->decodeToken($token, $publicCert);
+        } catch (InvalidValidationException $e) {
+            return $e->getCode();
         }
 
         if ($scopes === null) {
             return ValidationResults::Success;
-        } elseif (isset($decoded['scopeCount'])) {
+        } elseif (isset($decodedToken['scopeCount'])) {
             // user scopes not included within
             // the JWT as there are too many
             throw new ScopesNotDefinedException();
         }
 
-        $isSu = in_array('su', $decoded['scopes'], true);
-        $hasScope = count(array_intersect($scopes, $decoded['scopes'])) > 0;
+        $isSu = in_array('su', $decodedToken['scopes'], true);
+        $hasScope = count(array_intersect($scopes, $decodedToken['scopes'])) > 0;
 
         return $isSu || $hasScope
             ? ValidationResults::Success
             : ValidationResults::Unauthorised;
     }
 
+    /**
+     * Validate and decode a JWT token
+     *
+     * @param string $token a token to validate explicitly, if you do not
+     *      specify one the method tries to find one
+     * @param string $publicCert public key to validate the token
+     * @return array decoded token
+     *
+     * @throws DomainException invalid public key
+     * @throws InvalidArgumentException empty public key
+     * @throws UnexpectedValueException invalid token
+     */
+    protected function decodeToken($token, $rawPublicCert)
+    {
+        try {
+            // JWT::decode calls openssl_verify which will cause a fatal error
+            // if the certificate is invalid. Calling openssl_pkey_get_public
+            // first ensures that the certificate is valid before progressing.
+            if ($pubCert = openssl_pkey_get_public($rawPublicCert)) {
+                return (array) JWT::decode($token, $pubCert, ['RS256']);
+            }
+
+            throw new \InvalidArgumentException('cannot parse public key');
+        } catch (\DomainException $exception) {
+            $this->getLogger()->error('Invalid signature', [$exception]);
+            throw new InvalidValidationException(
```

Comment:
(In my opinion, we should be moving to more strictly-typed checked exceptions, rather than providing generic exceptions with things we need to inspect.  It's the way exceptions are meant to be used.)